nucypher/utilities/versioning.py

```diff
@@ -0,0 +1,83 @@
+"""
+ This file is part of nucypher.
+
+ nucypher is free software: you can redistribute it and/or modify
+ it under the terms of the GNU Affero General Public License as published by
+ the Free Software Foundation, either version 3 of the License, or
+ (at your option) any later version.
+
+ nucypher is distributed in the hope that it will be useful,
+ but WITHOUT ANY WARRANTY; without even the implied warranty of
+ MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ GNU Affero General Public License for more details.
+
+ You should have received a copy of the GNU Affero General Public License
+ along with nucypher.  If not, see <https://www.gnu.org/licenses/>.
+"""
+
+from abc import abstractmethod, ABC
+from typing import Dict
+
+
+class Versioned(ABC):
+
+    _BRAND_LENGTH = 2  # bytes
+    _VERSION_LENGTH = 2
+    _HEADER_SIZE = _BRAND_LENGTH + _VERSION_LENGTH
+
+    @classmethod
+    def from_bytes(cls, data):
+
+        # Metadata
+        brand = data[:cls._BRAND_LENGTH]
+        version_index = cls._BRAND_LENGTH + cls._VERSION_LENGTH
+        version_data = data[cls._BRAND_LENGTH:version_index]
+        version_number = int.from_bytes(version_data, 'big')
+
+        # Data passed to deserializer
+        remainder = data[version_index:]
+
+        # Validate and Deserialize
+        if brand != cls._brand():
+            raise ValueError(f"Incorrect brand.  Expected {cls._brand()}, Got {brand}")
+        if version_number == cls._version():
+            return cls._from_bytes_current(remainder)
+
+        handlers = cls._old_version_handlers()
+        try:
+            return handlers[version_number](remainder)
+        except KeyError:
+            raise ValueError(f"Incorrect or unknown version number ({version_number}).")
+
+    def __bytes__(self):
+        return self._header() + self._payload()
+
+    @classmethod
+    def _header(cls) -> bytes:
+        version_bytes = cls._version().to_bytes(cls._VERSION_LENGTH, 'big')
+        return cls._brand() + version_bytes
+
+    @classmethod
+    @abstractmethod
+    def _brand(cls) -> bytes:
+        raise NotImplementedError
+
+    @classmethod
+    @abstractmethod
+    def _version(cls) -> int:
+        raise NotImplementedError
+
+    @abstractmethod
+    def _payload(self) -> bytes:
+        """Returns the unversioned bytes serialized representation of this instance."""
+        raise NotImplementedError
+
+    @classmethod
+    @abstractmethod
+    def _old_version_handlers(cls) -> Dict:
```

Comment:
Thanks for the feedback!

`Versioned._old_version_handlers` is an internal method.  We need a way to compose the old version handlers on each subclass.  Moving the handler lookup logic inside this method means repeating it in each subclass's implementation or at least introducing an additional lookup method.

> instead of a Dict with all of the handlers for all of the previous versions?

There will still need to be a mapping somewhere, right?  Where do you suggest relocating it to?

> A Dict for the entire version history does not have to be generated or returned for every call to this function

I agree this feels somewhat imperfect. however the generation of this dictionary must have the method handlers in scope, and using an abstract method is useful here.

> This function can return None if no handler exists for the specified version, and therefore a KeyError does not have to be handled on L49.

I don't know If this as an advantage 😄

Reply:
> There will still need to be a mapping somewhere, right?

Not necessarily since within the function it could use a mapping (although I don't think it necessarily has to be), but it could also just be if-statements based on the version number?

> I agree this feels somewhat imperfect. however the generation of this dictionary must have the method handlers in scope, and using an abstract method is useful here.

It may help me to understand what an implementation of an old version handler would look like - I assumed it would probably be a reference to some Callable - can it be something else? Is scope an issue if presumably, these callables would be class methods or static methods (?) because `_old_version_hander(cls, version)` would be a class method?

Apologies if I am missing something obvious.

> I don't know If this as an advantage 😄

Haha fair - that one is more personal preference. I rather not have to catch `KeyError`. 😅

Reply:
> Not necessarily since within the function it could use a mapping (although I don't think it necessarily has to be), but it could also just be if-statements based on the version number?

This is essentially just avoiding the creation of an actual dictionary.  I call premature optimization! 😝 

> It may help me to understand what an implementation of an old version handler would look like 

Have a look here:
<https://gist.github.com/fjarri/bad73af26927e6fb3b1aa5f61f8863e3#file-versioning-py-L61>

> Haha fair - that one is more personal preference. I rather not have to catch KeyError. sweat_smile

As the old python adages go: "Better to ask for forgiveness than permission." and "Don't look before you leap!"

Reply:
Ended up removing the `try` in favor of membership checking a subsequent refactor 😆 .